loopy/translation_unit.py

```diff
@@ -770,19 +749,32 @@ def _collective_transform(*args, **kwargs):
 
 
 def update_table(callables_table, clbl_id, clbl):
+    """
+    Returns a tuple ``new_clbl_id, new_callables_table`` where
+    *new_callables_table* is a copy of *callables_table* with *clbl* in its
+    namespace. *clbl* is referred in *new_callables_table*'s namespace by
+    *new_clbl_id*.
+    """
     from loopy.kernel.function_interface import InKernelCallable
     assert isinstance(clbl, InKernelCallable)
 
     for i, c in callables_table.items():
         if c == clbl:
             return i, callables_table
 
-    while clbl_id in callables_table:
-        clbl_id = next_indexed_function_id(clbl_id)
+    if isinstance(clbl_id, ReductionOpFunction):
+        new_clbl_id = clbl_id.copy()
+    else:
+        assert isinstance(clbl_id, str)
+        from pytools import UniqueNameGenerator
+        all_substs = get_all_subst_names(callables_table)
+        ung = UniqueNameGenerator(set(callables_table.keys()) | all_substs)
```

Comment:
I think it would be good to factor this out, like `make_callable_unique_name_generator`.

Reply:
Yep, see bfd66f0.